Add basic structure for the controllers library and an implementation of an impedance controller

buschbapti opened the pull request:
This PR includes the implementation of the basic controllers structure and an impedance controller. The base Controller is template to accept both JointState and CartesianState controllers. The impedance controller is implemented in both Joint and Cartesian spaces.

For now, the test do not perform any checking. I just used it ti verify the syntax and base behavior of the impedance controller. We need to figure out better way of testing those.

This PR is still a draft as it requires the refactoring from #18 and #15. Once those are merged I will open this one officially.

buschbapti commented:
Now that all needed refactor branches have been merged I opened this PR. I have rebase it such that it includes only the files and commits that concern the controllers library

eeberhard commented on source/controllers/include/controllers/Controller.hpp:
```diff
@@ -0,0 +1,72 @@
+  /**
+   * @brief Compute the command based on the desired state and a feedback state
+   * To be redefined based on the actual controller implementation.
+   * This function accept any number of extra arguments.
```
No ellipsis in function declaration, so either this comment line is wrong or you meant to allow variadic argument below

buschbapti replied:
Yes I was not sure with that. I think we actually need an example where it makes sense to have those variadic arguments and at the end I could not find so I removed it in this declaration but forgot the comment.

eeberhard commented on source/controllers/include/controllers/Controller.hpp:
```diff
@@ -0,0 +1,72 @@
+};
+
+template <class SIn, class SOut>
+Controller<SIn, SOut>::Controller() {}
```
Is an explicit constructor necessary because of the templating? Otherwise I would remove the definition and declaration of the constructor if it's just the default constructor.

buschbapti replied:
Good question I actually don't think so. Except if we want to specify what are the accepted combinations (see comment below) then we can do that in the source file.

eeberhard commented on source/controllers/include/controllers/Controller.hpp:
```diff
@@ -0,0 +1,72 @@
+
+template <class SIn, class SOut>
+SOut Controller<SIn, SOut>::compute_command(const SIn&, ...) const {
+  throw exceptions::NotImplementedException("compute_command(const SIn& state, ...) not implemented for the base controller class");
```
Just wondering, why do you prefer this pattern with "not implemented" over using pure virtual functions? I guess so that you can actually compile code / test code with the base class or other still abstract derived classes?

buschbapti replied:
Mainly because I am not sure that a derived class not to redefine all the possibilities. If those are virtual pure then it become compulsory and if they are not defined in the base class there is an error at linkage

eeberhard commented on source/controllers/include/controllers/Controller.hpp:
```diff
@@ -0,0 +1,72 @@
+   * @param state the input state of the system. This function accept any number of extra arguments.
+   * @return the output command at the input state
+   */
+  virtual SOut compute_command(const SIn& state, ...) const;
```
I wonder if a non-const version of compute_command would ever be needed, for a controller that has some history dependence (such as an integrating error term or something like that)

A derived class would not be able to override this with a non-const definition, since the signature is different, and so you also can't access that method of a derived object through a base pointer.

I guess for that you could add an `update()` method too, but that may complicate the pattern when derived classes have unique public functions. Some libraries use a pattern of `c_`, `nc_` prefix or suffix to denote const, non-const, and other return type variants for functions. Maybe there can be a `compute_command_nc` for that special case.

But anyway, it's a bit of a specific point and we don't need it just yet. Just something to think about.

eeberhard replied:
Ah, I thought of a very pertinent example:
the "passive" part of the passive DS controller is indeed non-const. It needs a time delta, and accordingly adapts its energy tank based on state.

buschbapti replied:
Yes totally agree with that. I actually had to remove the const for PassiveDS implementation. I guess we can probably make the two versions of the functions. Which also explain why I am not going with pure virtual in the `Controller` base class as I assumed that some derived class might not need to redefine all of the variations of the `compute_command` function.

eeberhard commented on source/controllers/include/controllers/impedance/Impedance.hpp:
```diff
@@ -0,0 +1,133 @@
+};
+
+template <class S>
+Impedance<S>::Impedance(const Eigen::MatrixXd& stiffness, const Eigen::MatrixXd& damping, const Eigen::MatrixXd& inertia) : stiffness_(std::make_shared<StateRepresentation::Parameter<Eigen::MatrixXd>>("stiffness", stiffness)),
```
Can the constructor validate the template type and throw an unsupported exception if not joint state / cartesian state? Or will this instead be handled by base "not implemented" exception? (if so that's a pretty interesting consequence for using that over pure virtual functions - you can always compile but might fail at runtime)

buschbapti replied:
There is one possibility to do that is to define all the accepted implementations in the `.cpp` file. This way, the compiler will fail to compile with anything else. It is a bit tedious if you have many accepted classes (hello `Parameters`) but quite robust.

eeberhard commented on source/controllers/tests/testImpedanceController.cpp:
```diff
@@ -0,0 +1,51 @@
+  desired_state.set_linear_velocity(Eigen::Vector3d(1, 0, 0));
+  // set up a real state
+  CartesianState feedback_state("test");
+  feedback_state.set_orientation(Eigen::Quaterniond(0, 0, 1, 1));
```
sort of unrelated question - is the orientation quaternion regularly normalised in any of the state operations? Because at the moment this is of course not a unit quaternion. I would assume it is normalised at the appropriate step, but it's worth checking those assumptions!

buschbapti replied:
Quaternions are indeed normalized at each `set_orientation` call. This is why I am not bothering putting a unit quaternion.

eeberhard commented on source/controllers/include/controllers/impedance/Impedance.hpp:
```diff
@@ -0,0 +1,133 @@
+/**
```
I generally prefer attribution to remain in source control rather than the source itself, but if you want to keep it there I won't stop you

buschbapti replied:
Right yes make sense. I never know if those are necessary and where they should go.

eeberhard reviewed (approved):
Seems pretty good to me! Great start on this. I've got some comments, but mainly I approve.

I also know that the build structure and test structure will be expanded in the next stages, so I haven't made any comments there yet.

buschbapti commented:
I have addressed your comments. As you can see, we can restrict the Impedance to the specifics template definition. I will not do it in Controller as there will be too many combinations otherwise. Also adding non const version of the functions is also working.

domire8 reviewed (approved):
I don't have a lot of experience with templated stuff, but I trust you ;)

buschbapti commented:
As we discussed @eeberhard const functions have been removed to avoid troubles when deciding which function to use.

buschbapti commented:
I will now merge this then
